### PR TITLE
Feature/asv 639 ab testing implementation

### PR DESCRIPTION
**Pull request by paquiquinho**
**What does this PR do?**

This PR aims to implement an extensible and reusable AB Testing feature; This involved creating everything from the service to the interactor;

**Database changed?**

   Yes 

**Where should the reviewer start?**

- [x] AppViewPresenter.java
- [x] ABTestManager.java
- [x] ABTestCenterRepository.java

**How should this be manually tested?**

You'll need to check the if, on install press, the requests to assignment, impression, and, on click on continue/share, action; You should also verify the local cache by clicking install again without closing the app (you should see the experiment, make impression and action clicks but not the assignment) and the persistence by killing the app and pressing install again (it should act the same as with the local cache in this instance);

**What are the relevant tickets?**

  Tickets related to this pull-request: [ASV-639](<https://aptoide.atlassian.net/browse/ASV-639>)

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Partners build
- [ ] Unit tests pass
- [ ] Functional QA tests pass

### CODE REVIEW

File: app/src/main/java/cm/aptoide/pt/abtesting/ABTestManager.java [hunk 1]
```diff
@@ -1,116 +1,43 @@
...
-          return test;
-        }
-      }
-      throw new IllegalArgumentException("No AB test for name: " + name);
-    } else {
-      return getControl(name);
-    }
-  }
+  public enum ExperimentType {
+    SHARE_DIALOG("android_implement"), BUTTON_COLOR("dumb_test");
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/abtesting/ABTestManager.java]: Should this be here? "dumb_test"

File: app/src/main/java/cm/aptoide/pt/abtesting/Experiment.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
+package cm.aptoide.pt.abtesting;
+
+/**
+ * Created by franciscocalado on 15/06/18.
+ */
+
+public class Experiment {
+  private static final long TWENTY_FOUR_HOURS = 864000000;
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/abtesting/Experiment.java]: This is 240h right?

File: aptoide-database/src/main/java/cm/aptoide/pt/database/schedulers/RealmSchedulers.java [hunk 3]
```diff
@@ -37,7 +37,7 @@ public static Scheduler getScheduler() {
     return getInstance().threadScheduler;
   }
 
-  /**
+  /**Å
```
Comment (jdandrade) [inline aptoide-database/src/main/java/cm/aptoide/pt/database/schedulers/RealmSchedulers.java]: Probably typo :p

File: app/src/vanilla/java/cm/aptoide/pt/app/view/AppViewPresenter.java [hunk 4]
```diff
@@ -732,13 +761,27 @@ private void handleInstallButtonClick() {
...
+        .subscribe(__ -> {
+        }, err -> crashReport.log(err));
+  }
+
+  private void showRecommendsDialog(boolean isLoggedIn, String packageName, Experiment experiment) {
     if (isLoggedIn && appViewManager.shouldShowRecommendsPreviewDialog()) {
-      view.showRecommendsDialog();
+      view.showRecommendsDialog(experiment);
       appViewAnalytics.sendLoggedInRecommendAppDialogShowEvent(packageName);
+      dialogImpression.onNext(true);
```
Comment (jdandrade) [inline app/src/vanilla/java/cm/aptoide/pt/app/view/AppViewPresenter.java]: We shouldn't use the publish subject here to "communicate" the impression. I think we can turn this method an observable and make the impression request in the chain that calls this.

File: app/src/main/java/cm/aptoide/pt/abtesting/ABTestActionBody.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
+package cm.aptoide.pt.abtesting;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by franciscocalado on 19/06/18.
+ */
+
+public class ABTestActionBody {
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/abtesting/ABTestActionBody.java]: Isn't this and ABTestImpressionBody the same structure? Could we merge this someway and just inject or set the differences (which it seems like it is only the assignment name)?

File: app/src/main/java/cm/aptoide/pt/abtesting/Experiment.java (hunk 2 above)
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/abtesting/Experiment.java]: Fixed

File: app/src/main/java/cm/aptoide/pt/abtesting/ABTestManager.java (hunk 1 above)
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/abtesting/ABTestManager.java]: Fixed

File: aptoide-database/src/main/java/cm/aptoide/pt/database/schedulers/RealmSchedulers.java (hunk 3 above)
Comment (paquiquinho) [reply aptoide-database/src/main/java/cm/aptoide/pt/database/schedulers/RealmSchedulers.java]: Fixed

File: app/src/main/java/cm/aptoide/pt/abtesting/ABTestActionBody.java (hunk 5 above)
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/abtesting/ABTestActionBody.java]: Fixed

File: app/src/vanilla/java/cm/aptoide/pt/app/view/AppViewPresenter.java (hunk 4 above)
Comment (paquiquinho) [reply app/src/vanilla/java/cm/aptoide/pt/app/view/AppViewPresenter.java]: Not really fixed but was discussed

**Pull request merged**